components/movementsensor/wheeledodometry/wheeledodometry.go

```diff
@@ -0,0 +1,330 @@
+}
+
+func (o *odometry) LinearAcceleration(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	return r3.Vector{}, movementsensor.ErrMethodUnimplementedAngularVelocity
+}
+
+func (o *odometry) Orientation(ctx context.Context, extra map[string]interface{}) (spatialmath.Orientation, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	ov := &spatialmath.OrientationVector{Theta: o.orientation.Yaw, OX: 0, OY: 0, OZ: 1}
+	return ov, nil
+}
+
+func (o *odometry) CompassHeading(ctx context.Context, extra map[string]interface{}) (float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return 0, movementsensor.ErrMethodUnimplementedCompassHeading
+}
+
+func (o *odometry) LinearVelocity(ctx context.Context, extra map[string]interface{}) (r3.Vector, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return o.linearVelocity, nil
+}
+
+func (o *odometry) Position(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	return geo.NewPoint(o.position.X, o.position.Y), o.position.Z, nil
+}
+
+func (o *odometry) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	return movementsensor.Readings(ctx, o, extra)
+}
+
+func (o *odometry) Accuracy(ctx context.Context, extra map[string]interface{}) (map[string]float32, error) {
+	return map[string]float32{}, movementsensor.ErrMethodUnimplementedAccuracy
+}
+
+func (o *odometry) Properties(ctx context.Context, extra map[string]interface{}) (*movementsensor.Properties, error) {
+	return &movementsensor.Properties{
+		LinearVelocitySupported:  true,
+		AngularVelocitySupported: true,
+		OrientationSupported:     true,
+		PositionSupported:        true,
+	}, nil
+}
+
+func (o *odometry) Close(ctx context.Context) error {
+	o.cancelFunc()
+	o.activeBackgroundWorkers.Wait()
+	return nil
+}
+
+// trackPosition uses the motor positions to calculate an estimation of the position, orientation,
+// linear velocity, and angular velocity of the wheeled base.
+// The estimations in this function are based on the math outlined in this article:
+// https://stuff.mit.edu/afs/athena/course/6/6.186/OldFiles/2005/doc/odomtutorial/odomtutorial.pdf
+func (o *odometry) trackPosition(ctx context.Context) {
+	o.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer o.activeBackgroundWorkers.Done()
+		ticker := time.NewTicker(time.Duration(o.timeIntervalMSecs) * time.Millisecond)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Sleep until it's time to update the position again.
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			// Use GetInParallel to ensure the left and right motors are polled at the same time.
+			positionFuncs := func() []rdkutils.FloatFunc {
+				fs := []rdkutils.FloatFunc{}
+
+				// These reads of the left and right motors can race with `Reconfigure`.
+				o.mu.Lock()
+				defer o.mu.Unlock()
```

Comment:
Wait, comment doesn't match code! These don't lock the mutex when we read the motors. This code is currently:

1. Lock the mutex
2. Put functions in an orientation
3. Unlock the mutex
4. Run the functions in parallel, which involves interacting with `o.motors`

Try putting the mutex-locking around the call to `GetInParallel` instead.

Reply:
actually, I made a mistake here. I don't think I even need the lock and unlock because in the original use of this, the code is looping through all left and right motors. so it might become relevant at some point, but as of right now I'm only reading one left and one right so it should be fine I'd think? This is how `RunInParallel` is used in wheeled base
```
goForFuncs := func() []rdkutils.SimpleFunc {
  ret := []rdkutils.SimpleFunc{}
  
  // These reads of `wb.left` and `wb.right` can race with `Reconfigure`.
  wb.mu.Lock()
  defer wb.mu.Unlock()
  for _, m := range wb.left {
	  motor := m
	  ret = append(ret, func(ctx context.Context) error { return motor.GoFor(ctx, leftRPM, leftRotations, nil) })
  }
  
  for _, m := range wb.right {
	  motor := m
	  ret = append(ret, func(ctx context.Context) error { return motor.GoFor(ctx, rightRPM, rightRotations, nil) })
  }
  return ret
}()

Reply:
going to merge so I can open my next PR but I can add changes about this to that other PR if it becomes an issue.